### PR TITLE
BREAKING CHANGE: refactor into libraries. add join flag. support multiple pr jobs

### PR DESCRIPTION
**Pull request by petey**
Context:
========
Planning ahead to add workflow cycle detection, but wanted to get api changes out sooner.

Objective:
----------
* Mark edges that are part of a join with a flag in the edge node.
* Add node objects for ~commit and ~pr events
* BREAKING CHANGE: changed signature of the all methods to be more consistent.

### CODE REVIEW

File: lib/getNextJobs.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+'use strict';
+
+/**
+ * Calculate the next jobs to execute, given a workflow and a trigger job
+ * @method getNextJobs
+ * @param  {Object}    workflow         Directed graph representation of workflow
```
Comment (tkyi) [inline lib/getNextJobs.js]: Can you rename this variable to `workflowGraph` so we don't get it mixed up with the old concept of `workflow`?

File: lib/getNextJobs.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+    }
+
+    if (obj.trigger === '~pr' && !obj.prNum) {
+        throw new Error('Must provide a PR number with "~pr" trigger');
+    }
+
+    workflow.edges.forEach((edge) => {
+        if (edge.src === obj.trigger) {
+            // Make PR jobs PR-$num:$cloneJob (not sure how to better handle multiple PR jobs)
+            jobs.add(obj.trigger === '~pr' ? `PR-${obj.prNum}:${edge.dest}` : edge.dest);
```
Comment (tkyi) [inline lib/getNextJobs.js]: I think we're going with `PR-${obj.prNum}-${edge.dest}`

Comment (minz1027) [reply lib/getNextJobs.js]: I think we should use `:` cuz jobName can actually contains/start with `-`.

File: lib/getWorkflow.js [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        throw new Error('No Job config provided');
+    }
+
+    const hasRequiresConfig = Object.keys(jobConfig)
+        .some(j => Array.isArray(jobConfig[j].requires));
+
+    if (useLegacy && !hasRequiresConfig) {
+        // Work out whether there is a user defined workflow,
+        // or if we use the order of jobs defined in jobConfig
+        const workflow = Array.isArray(pipelineConfig.workflow) ?
```
Comment (minz1027) [inline lib/getWorkflow.js]: When it goes through the config parser, if will add `main` to the workflow.  And workflow saved in the pipeline will always have that main job in workflow.
https://api.screwdriver.cd/v4/pipelines

Can we have some check for `pipelineConfig.workflow`? If it has main then take it out

Comment (petey) [reply lib/getWorkflow.js]: fixed

File: README.md [hunk 4]
```diff
@@ -9,6 +9,26 @@
 npm install screwdriver-workflow-parser
 ```
 
+```
+const { getWorkflow, getNextJobs } = require('screwdriver-workflow-parser');
+
+// Calculate the directed graph workflow from a pipeline config (and parse legacy workflows)
+const workflowGraph = getWorkflow({ config: pipelineConfig, useLegacy: true });
```
Comment (tkyi) [inline README.md]: need to update examples here

File: lib/getNextJobs.js [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * Calculate the next jobs to execute, given a workflow and a trigger job
+ * @method getNextJobs
+ * @param  {Object}    workflowGraph    Directed graph representation of workflow
+ * @param  {Object}    obj
+ * @param  {String}    obj.trigger      The triggering event (~pr, ~commit, jobName)
+ * @param  {String}    [obj.prNum]      The PR number (required when ~pr trigger)
+ * @return {Array}                      List of job names
+ */
+const getNextJobs = (workflowGraph, obj) => {
```
Comment (tkyi) [inline lib/getNextJobs.js]: probably want to use "config" to be consistent

File: lib/getWorkflow.js [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        }
+    });
+
+    return edges;
+};
+
+/**
+ * Given a pipeline config, return a directed graph configuration that describes the workflow
+ * @method getWorkflow
+ * @param  {Object}    pipeline                 A Pipeline Config
```
Comment (tkyi) [inline lib/getWorkflow.js]: Need to update JSdocs to match new arg name (`pipelineConfig`)

**Pull request merged**